devito/passes/iet/dtypes.py

```diff
@@ -2,8 +2,44 @@
 import ctypes
 
 from devito.ir import FindSymbols, Uxreplace
+from devito.ir.iet.nodes import Dereference
+from devito.tools.utils import as_tuple
+from devito.types.basic import Symbol
 
-__all__ = ['lower_complex']
+__all__ = ['lower_scalar_half', 'lower_complex']
+
+
+def lower_scalar_half(iet, lang, sregistry):
+    """
+    Lower half float scalars to pointers (special case, since we can't
+    pass them directly for lack of a ctypes equivalent)
+    """
+    if lang.get('half_types') is None:
+        return iet, {}
+
+    # dtype mappings for float16
+    half, half_p = lang['half_types']
+
+    body = []  # derefs to prepend to the body
+    body_mapper = {}
+    params_mapper = {}
+
+    for s in FindSymbols('scalars').visit(iet):
+        if s.dtype != np.float16 or s not in iet.parameters:
+            continue
+
+        ptr = s._rebuild(dtype=half_p)
+        val = Symbol(name=sregistry.make_name(prefix='hf'), dtype=half, is_const=True)
+
+        params_mapper[s] = ptr
+        body_mapper[s] = val
+        body.append(Dereference(val, ptr))  # val = *ptr
+
+    body.extend(as_tuple(Uxreplace(body_mapper).visit(iet.body)))
```

Comment:
why does it. need extend? `Uxreplace(body_mapper).visit(iet.body)` should doo all it needs no?

Reply:
Well this prepends whatever I've put in `body` within the loop (those deref expressions). What would be a better way to do this? I can just append to `iet.body` I guess and do the mapping on the whole thing, including what I've added